File: backend/app/api/v1/documents.py
"""Document REST endpoints — upload, download, preview, CRUD."""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, Query, Response, UploadFile, status
from fastapi.responses import Response as FastAPIResponse
from sqlalchemy.orm import Session

from app.core.exceptions import NotFoundError
from app.dependencies.db import get_db
from app.schemas.common import DataResponse, ListResponse, PaginationMeta
from app.schemas.document import DocumentCreate, DocumentRead, DocumentUpdate
from app.services.document import DocumentService

router = APIRouter(tags=["documents"])


@router.get(
    "/documents",
    response_model=ListResponse[DocumentRead],
)
def list_documents(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=25, ge=1, le=100),
    asset_id: UUID | None = None,
    category: str | None = None,
    search: str | None = Query(default=None, max_length=200),
    db: Session = Depends(get_db),
) -> ListResponse[DocumentRead]:
    service = DocumentService(db)
    items, total = service.list_all(
        page=page,
        limit=limit,
        asset_id=asset_id,
        category=category,
        search=search,
    )
    return ListResponse(
        data=[service.to_read(item) for item in items],
        pagination=PaginationMeta.from_totals(page=page, limit=limit, total=total),
    )


@router.get(
    "/assets/{asset_id}/documents",
    response_model=ListResponse[DocumentRead],
)
def list_asset_documents(
    asset_id: UUID,
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=25, ge=1, le=100),
    category: str | None = None,
    db: Session = Depends(get_db),
) -> ListResponse[DocumentRead]:
    service = DocumentService(db)
    items, total = service.list_for_asset(
        asset_id,
        page=page,
        limit=limit,
        category=category,
    )
    return ListResponse(
        data=[service.to_read(item) for item in items],
        pagination=PaginationMeta.from_totals(page=page, limit=limit, total=total),
    )


@router.post(
    "/documents/upload",
    response_model=DataResponse[DocumentRead],
    status_code=status.HTTP_201_CREATED,
)
async def upload_document(
    asset_id: UUID = Form(...),
    file: UploadFile = File(...),
    name: str | None = Form(default=None),
    category: str | None = Form(default=None),
    notes: str | None = Form(default=None),
    db: Session = Depends(get_db),
) -> DataResponse[DocumentRead]:
    """Multipart upload — stores binary and creates document metadata."""
    data = await file.read()
    service = DocumentService(db)
    document = service.upload(
        asset_id=asset_id,
        filename=file.filename or "upload.bin",
        content_type=file.content_type,
        data=data,
        name=name,
        category=category,
        notes=notes,
    )
    return DataResponse(data=service.to_read(document))


@router.post(
    "/documents",
    response_model=DataResponse[DocumentRead],
    status_code=status.HTTP_201_CREATED,
)
def create_document(
    payload: DocumentCreate,
    db: Session = Depends(get_db),
) -> DataResponse[DocumentRead]:
    """Metadata-only create (no file bytes). Prefer /documents/upload."""
    service = DocumentService(db)
    document = service.create(payload)
    return DataResponse(data=service.to_read(document))


@router.get(
    "/documents/{document_id}",
    response_model=DataResponse[DocumentRead],
)
def get_document(
    document_id: UUID,
    db: Session = Depends(get_db),
) -> DataResponse[DocumentRead]:
    service = DocumentService(db)
    document = service.get(document_id)
    return DataResponse(data=service.to_read(document))


@router.get("/documents/{document_id}/download")
def download_document(
    document_id: UUID,
    db: Session = Depends(get_db),
) -> FastAPIResponse:
    service = DocumentService(db)
    document, data = service.read_file(document_id)
    headers = {
        "Content-Disposition": f'attachment; filename="{document.filename}"',
    }
    return FastAPIResponse(
        content=data,
        media_type=document.mime_type or "application/octet-stream",
        headers=headers,
    )


@router.get("/documents/{document_id}/preview")
def preview_document(
    document_id: UUID,
    db: Session = Depends(get_db),
) -> FastAPIResponse:
    """Inline file response for browser preview (PDF/images)."""
    service = DocumentService(db)
    document, data = service.read_file(document_id)
    headers = {
        "Content-Disposition": f'inline; filename="{document.filename}"',
    }
    return FastAPIResponse(
        content=data,
        media_type=document.mime_type or "application/octet-stream",
        headers=headers,
    )


@router.get("/documents/{document_id}/thumbnail")
def document_thumbnail(
    document_id: UUID,
    db: Session = Depends(get_db),
) -> FastAPIResponse:
    """Serve the generated thumbnail when available (Phase 9)."""
    service = DocumentService(db)
    document = service.get(document_id)
    if not document.thumbnail_path:
        raise NotFoundError(
            "THUMBNAIL_NOT_FOUND",
            "Thumbnail has not been generated for this document yet.",
        )
    try:
        data = service.storage.read(document.thumbnail_path)
    except FileNotFoundError as exc:
        raise NotFoundError(
            "THUMBNAIL_NOT_FOUND",
            "Thumbnail file is missing from disk.",
        ) from exc

    # Derivatives are jpeg/png/webp — sniff from path extension.
    path = document.thumbnail_path.lower()
    if path.endswith(".jpg") or path.endswith(".jpeg"):
        media = "image/jpeg"
    elif path.endswith(".webp"):
        media = "image/webp"
    else:
        media = "image/png"
    return FastAPIResponse(
        content=data,
        media_type=media,
        headers={"Content-Disposition": 'inline; filename="thumbnail"'},
    )


@router.patch(
    "/documents/{document_id}",
    response_model=DataResponse[DocumentRead],
)
def update_document(
    document_id: UUID,
    payload: DocumentUpdate,
    db: Session = Depends(get_db),
) -> DataResponse[DocumentRead]:
    service = DocumentService(db)
    document = service.update(document_id, payload)
    return DataResponse(data=service.to_read(document))


@router.delete(
    "/documents/{document_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_document(
    document_id: UUID,
    db: Session = Depends(get_db),
) -> Response:
    DocumentService(db).delete(document_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/actions/documents.ts
"use server";

import { revalidatePath } from "next/cache";

import type { Document } from "@/types/domain";

import { runAction, runListAction, withServerContext } from "@/lib/server/action-context";
import { requireRole } from "@/lib/server/authorize";
import { toDocument } from "@/lib/server/mappers";
import { parsePagination } from "@/lib/server/pagination";
import { DocumentService } from "@/lib/server/services/documents";

export type DocumentCreateInput = {
  asset_id: string;
  name: string;
  filename: string;
  mime_type?: string | null;
  size_bytes?: number | null;
  storage_path?: string | null;
  category?: string;
  notes?: string | null;
};

export type DocumentUpdateInput = {
  name?: string;
  notes?: string | null;
  category?: string;
};

const DOCUMENT_ROUTES = [
  "/dashboard/development",
  "/dashboard/database",
  "/dashboard/properties/[id]",
] as const;

function revalidateDocumentRoutes() {
  for (const path of DOCUMENT_ROUTES) revalidatePath(path);
}

export async function listDocumentsForAsset(
  assetId: string,
  params?: { page?: number; limit?: number; category?: string },
) {
  return runListAction<Document>(async () => {
    const { client } = await withServerContext();
    const { page, limit } = parsePagination(params?.page, params?.limit);
    const service = new DocumentService(client);
    const { items, total } = await service.listForAsset(assetId, {
      page,
      limit,
      category: params?.category ?? null,
    });
    return { items: items.map(toDocument), total, page, limit };
  });
}

export async function listDocuments(params?: {
  page?: number;
  limit?: number;
  asset_id?: string;
  category?: string;
  search?: string;
}) {
  return runListAction<Document>(async () => {
    const { client } = await withServerContext();
    const { page, limit } = parsePagination(params?.page, params?.limit);
    const service = new DocumentService(client);
    const { items, total } = await service.listAll({
      page,
      limit,
      asset_id: params?.asset_id ?? null,
      category: params?.category ?? null,
      search: params?.search ?? null,
    });
    return { items: items.map(toDocument), total, page, limit };
  });
}

export async function getDocument(id: string) {
  return runAction<Document>(async () => {
    const { client } = await withServerContext();
    const service = new DocumentService(client);
    return toDocument(await service.get(id));
  });
}

export async function createDocument(payload: DocumentCreateInput) {
  return runAction<Document>(async () => {
    const ctx = await withServerContext();
    const actor = requireRole(ctx.actor, "operator", "create", "document");
    const service = new DocumentService(ctx.client, { actor });
    const document = await service.create(payload);
    revalidateDocumentRoutes();
    return toDocument(document);
  });
}

export async function updateDocument(id: string, payload: DocumentUpdateInput) {
  return runAction<Document>(async () => {
    const ctx = await withServerContext();
    const actor = requireRole(ctx.actor, "operator", "update", "document");
    const service = new DocumentService(ctx.client, { actor });
    const document = await service.update(id, payload);
    revalidateDocumentRoutes();
    return toDocument(document);
  });
}

export async function deleteDocument(id: string) {
  return runAction<null>(async () => {
    const ctx = await withServerContext();
    const actor = requireRole(ctx.actor, "manager", "delete", "document");
    const service = new DocumentService(ctx.client, { actor });
    await service.delete(id);
    revalidateDocumentRoutes();
    return null;
  });
}

/** Upload a document binary from multipart form data. */
export async function uploadDocument(formData: FormData) {
  return runAction<Document>(async () => {
    const ctx = await withServerContext();
    const actor = requireRole(ctx.actor, "operator", "upload", "document");
    const assetId = formData.get("asset_id");
    const file = formData.get("file");
    if (typeof assetId !== "string" || !(file instanceof File)) {
      throw new Error("asset_id and file are required.");
    }
    const service = new DocumentService(ctx.client, { actor });
    const data = new Uint8Array(await file.arrayBuffer());
    const document = await service.upload({
      asset_id: assetId,
      filename: file.name,
      content_type: file.type || null,
      data,
      name: typeof formData.get("name") === "string" ? (formData.get("name") as string) : undefined,
      category: typeof formData.get("category") === "string" ? (formData.get("category") as string) : undefined,
      notes: typeof formData.get("notes") === "string" ? (formData.get("notes") as string) : undefined,
    });
    revalidateDocumentRoutes();
    return toDocument(document);
  });
}